File: node/src/agents/wif_agent.ts
/**
 * WifAgent — finance expert (Node.js port).
 * In-memory expense tracker + financial calculators. agentType: "financier"
 */

import { Actor, MqttPublisher } from "../core/actor";
import { Message, MessageType } from "../core/types";

interface Expense {
  amount:    number;
  category:  string;
  note:      string;
  timestamp: number;
}

const HELP = `**WIF — Finance Expert** 💹
_Waldiez In Finance_

\`\`\`
add <amount> [category] [note]          log an expense
budget <amount>                         set monthly budget limit
report                                  spending breakdown by category
balance                                 budget vs total spent
compound <principal> <rate%> <years>    compound interest (monthly)
loan <principal> <rate%> <months>       monthly loan payment
roi <investment> <return>               return on investment %
tax <income> [bracket%]                 simple tax estimate (default 25%)
tip <bill> [percent]                    tip calculator (default 15%)
help                                    this message
\`\`\``;

export class WifAgent extends Actor {
  private _expenses: Expense[] = [];
  private _budget = 0;

  constructor(publish: MqttPublisher, actorId?: string) {
    super("wif-agent", publish, actorId);
  }

  protected override async onStart(): Promise<void> {
    this.publishSpawn("financier");
  }

  override async handleMessage(msg: Message): Promise<void> {
    if (msg.type !== MessageType.Task && msg.type !== MessageType.Text) return;
    let text = Actor.extractText(msg.payload).trim();
    if (!text) return;
    text = Actor.stripPrefix(text, "@wif-agent", "@wif_agent");
    const parts = text.trim().split(/\s+/);
    const cmd = (parts[0] ?? "").toLowerCase();
    const args = parts.slice(1);
    const reply = this._dispatch(cmd, args);
    this.replyChat(reply);
    this.metrics.tasksCompleted++;
  }

  private _dispatch(cmd: string, args: string[]): string {
    switch (cmd) {
      case "add":      return this._cmdAdd(args);
      case "budget":   return this._cmdBudget(args);
      case "report":   return this._cmdReport();
      case "balance":  return this._cmdBalance();
      case "compound": return this._cmdCompound(args);
      case "loan":     return this._cmdLoan(args);
      case "roi":      return this._cmdRoi(args);
      case "tax":      return this._cmdTax(args);
      case "tip":      return this._cmdTip(args);
      case "help":
      case "":         return HELP;
      default:         return `Unknown command: \`${cmd}\`. Type \`help\`.`;
    }
  }

  private _cmdAdd(args: string[]): string {
    if (!args[0]) return "Usage: `add <amount> [category] [note]`";
    const amount = parseFloat(args[0].replace(/^[$€£¥+]/, ""));
    if (isNaN(amount) || amount <= 0) return `Invalid amount: \`${args[0]}\``;
    const category = args[1]?.toLowerCase() ?? "misc";
    const note = args.slice(2).join(" ");
    this._expenses.push({ amount, category, note, timestamp: Date.now() });
    const header = `Logged **$${amount.toFixed(2)}** → \`${category}\`` + (note ? ` — _${note}_` : "");
    if (this._budget > 0) {
      const spent = this._expenses.reduce((s, e) => s + e.amount, 0);
      const pct = Math.min((spent / this._budget) * 100, 999);
      const icon = pct >= 100 ? "🚨" : pct >= 80 ? "⚠" : "✓";
      return `${header}\n${icon} Budget: $${spent.toFixed(2)} / $${this._budget.toFixed(2)} (${pct.toFixed(0)}%)`;
    }
    return header;
  }

  private _cmdBudget(args: string[]): string {
    if (!args[0]) return "Usage: `budget <amount>`";
    const amount = parseFloat(args[0].replace(/^[$€£¥]/, ""));
    if (isNaN(amount)) return `Invalid amount: \`${args[0]}\``;
    this._budget = amount;
    const spent = this._expenses.reduce((s, e) => s + e.amount, 0);
    const pct = amount > 0 ? Math.min((spent / amount) * 100, 999) : 0;
    return `Budget set: **$${amount.toFixed(2)}** | Spent so far: $${spent.toFixed(2)} (${pct.toFixed(0)}%)`;
  }

  private _cmdReport(): string {
    if (this._expenses.length === 0) return "No expenses yet. Try: `add 25 food coffee`";
    const byCat: Record<string, number> = {};
    for (const e of this._expenses) byCat[e.category] = (byCat[e.category] ?? 0) + e.amount;
    const total = Object.values(byCat).reduce((s, v) => s + v, 0);
    const rows = Object.entries(byCat)
      .sort(([, a], [, b]) => b - a)
      .map(([cat, amt]) => `  **${cat}**: $${amt.toFixed(2)} (${((amt / total) * 100).toFixed(0)}%)`);
    return `**Expense Report** (${this._expenses.length} transactions)\n\n${rows.join("\n")}\n\n**Total: $${total.toFixed(2)}**`;
  }

  private _cmdBalance(): string {
    const spent = this._expenses.reduce((s, e) => s + e.amount, 0);
    if (this._budget <= 0) {
      return `**Balance**\n\nTotal spent: **$${spent.toFixed(2)}** (${this._expenses.length} transactions)\nBudget: _not set_`;
    }
    const rem = this._budget - spent;
    const pct = Math.min((spent / this._budget) * 100, 999);
    const icon = pct >= 100 ? "🚨" : pct >= 80 ? "⚠" : "✓";
    return (
      `**Monthly Budget Balance**\n\n${icon} $${spent.toFixed(2)} / $${this._budget.toFixed(2)} (${pct.toFixed(0)}%)\n` +
      (rem >= 0 ? `$${rem.toFixed(2)} left` : `$${(-rem).toFixed(2)} over budget`)
    );
  }

  private _cmdCompound(args: string[]): string {
    if (args.length < 3) return "Usage: `compound <principal> <rate%> <years>`";
    const [p, r, t] = args.map((a, i) => parseFloat(i === 1 ? a.replace(/%$/, "") : a.replace(/^\$/, "")));
    if ([p, r, t].some(isNaN)) return "Invalid numbers.";
    const n  = 12;
    const fv = p * Math.pow(1 + r / 100 / n, n * t);
    return `**Compound Interest**\n\nPrincipal: $${p.toLocaleString("en")} | Rate: ${r}% | Term: ${t}y\n→ **Future Value: $${fv.toLocaleString("en", { minimumFractionDigits: 2, maximumFractionDigits: 2 })}** (interest: $${(fv - p).toLocaleString("en", { minimumFractionDigits: 2, maximumFractionDigits: 2 })})`;
  }

  private _cmdLoan(args: string[]): string {
    if (args.length < 3) return "Usage: `loan <principal> <rate%> <months>`";
    const p      = parseFloat(args[0].replace(/^\$/, ""));
    const annual = parseFloat(args[1].replace(/%$/, ""));
    const n      = parseFloat(args[2]);
    if ([p, annual, n].some(isNaN)) return "Invalid numbers.";
    const r = annual / 100 / 12;
    const monthly = r ? p * r * Math.pow(1 + r, n) / (Math.pow(1 + r, n) - 1) : p / n;
    const total   = monthly * n;
    return `**Loan Calculator**\n\nPrincipal: $${p.toLocaleString("en")} | Rate: ${annual}% | Term: ${n}mo\n→ **Monthly: $${monthly.toFixed(2)}** | Total: $${total.toFixed(2)} | Interest: $${(total - p).toFixed(2)}`;
  }

  private _cmdRoi(args: string[]): string {
    if (args.length < 2) return "Usage: `roi <investment> <return>`";
    const inv = parseFloat(args[0].replace(/^\$/, ""));
    const ret = parseFloat(args[1].replace(/^\$/, ""));
    if ([inv, ret].some(isNaN) || inv === 0) return "Invalid numbers.";
    const gain = ret - inv;
    const roi  = (gain / inv) * 100;
    return `**ROI**\n\nInvestment: $${inv.toFixed(2)} | Return: $${ret.toFixed(2)}\n→ Gain: $${gain >= 0 ? "+" : ""}${gain.toFixed(2)} | **ROI: ${roi >= 0 ? "+" : ""}${roi.toFixed(2)}%**`;
  }

  private _cmdTax(args: string[]): string {
    if (!args[0]) return "Usage: `tax <income> [bracket%]`";
    const income = parseFloat(args[0].replace(/^\$/, ""));
    const rate   = args[1] ? parseFloat(args[1].replace(/%$/, "")) : 25;
    if ([income, rate].some(isNaN)) return "Invalid numbers.";
    const tax = income * rate / 100;
    return `**Tax Estimate**\n\nIncome: $${income.toLocaleString("en")} | Rate: ${rate}%\n→ Tax: **$${tax.toLocaleString("en", { minimumFractionDigits: 2 })}** | Net: **$${(income - tax).toLocaleString("en", { minimumFractionDigits: 2 })}**\n_Simplified estimate — consult a professional._`;
  }

  private _cmdTip(args: string[]): string {
    if (!args[0]) return "Usage: `tip <bill> [percent]`";
    const bill = parseFloat(args[0].replace(/^\$/, ""));
    const pct  = args[1] ? parseFloat(args[1].replace(/%$/, "")) : 15;
    if ([bill, pct].some(isNaN)) return "Invalid numbers.";
    const tip   = bill * pct / 100;
    const total = bill + tip;
    const splits = [1, 2, 3, 4, 5].map((n) => `  ${n} people: $${(total / n).toFixed(2)}/person`).join("\n");
    return `**Tip Calculator**\n\nBill: $${bill.toFixed(2)} | Tip (${pct.toFixed(0)}%): $${tip.toFixed(2)} | **Total: $${total.toFixed(2)}**\n\n${splits}`;
  }
}

File: agentflow/agents/wif_agent.py
"""
WifAgent — Finance expert (WIF = Waldiez In Finance).

In-memory finance tracker with calculations. No external API key required.
All state persisted via Actor pickle mechanism.

Commands (prefix @wif-agent stripped automatically):
  add <amount> [category] [note]          log an expense
  budget <amount>                         set monthly budget
  report                                  breakdown by category
  balance                                 budget vs total spent
  compound <principal> <rate%> <years>    compound interest (monthly)
  loan <principal> <rate%> <months>       monthly loan payment
  roi <investment> <return>               ROI percentage
  tax <income> [bracket%]                 simple tax estimate (default 25%)
  tip <bill> [percent]                    tip calculator (default 15%)
  help                                    list all commands
"""

from __future__ import annotations

import logging
import time
from typing import Any

from ..core.actor import Actor, Message, MessageType

logger = logging.getLogger(__name__)

_HELP = """\
**WIF — Finance Expert** 💹
_Waldiez In Finance_

```
add <amount> [category] [note]          log an expense
budget <amount>                         set monthly budget limit
report                                  spending breakdown by category
balance                                 budget vs total spent
compound <principal> <rate%> <years>    compound interest (monthly)
loan <principal> <rate%> <months>       monthly loan payment
roi <investment> <return>               return on investment %
tax <income> [bracket%]                 simple tax estimate (default 25%)
tip <bill> [percent]                    tip calculator (default 15%)
help                                    this message
```"""


class WifAgent(Actor):
    """Finance-expert actor — no API key, all calculations local."""

    def __init__(self, **kwargs: Any) -> None:
        kwargs.setdefault("name", "wif-agent")
        super().__init__(**kwargs)
        self.protected  = False
        self._expenses: list[dict] = []
        self._budget: float        = 0.0

    async def on_start(self) -> None:
        self._expenses = self.recall("expenses", [])
        self._budget   = float(self.recall("budget", 0.0))
        await self._mqtt_publish(
            f"agents/{self.actor_id}/spawn",
            {"agentId": self.actor_id, "agentName": self.name, "agentType": "financier", "timestamp": time.time()},
        )
        logger.info("[%s] started — %d expenses loaded", self.name, len(self._expenses))

    async def on_stop(self) -> None:
        self._save()

    def _save(self) -> None:
        self.persist("expenses", self._expenses)
        self.persist("budget", self._budget)

    async def handle_message(self, msg: Message) -> None:
        if msg.type not in (MessageType.TASK, MessageType.RESULT):
            return
        payload = msg.payload or {}
        text = str(
            payload.get("text") or payload.get("content") or payload.get("task") or ""
            if isinstance(payload, dict) else payload
        ).strip()
        if not text:
            return
        reply = self._dispatch(text)
        self._save()
        await self._reply(reply)

    async def _reply(self, content: str) -> None:
        await self._mqtt_publish(
            f"agents/{self.actor_id}/chat",
            {"from": self.name, "to": "user", "content": content, "timestamp": time.time()},
        )

    def _dispatch(self, raw: str) -> str:
        text = raw
        for pfx in ("@wif-agent", "@wif_agent"):
            if text.lower().startswith(pfx):
                text = text[len(pfx):].lstrip()
                break
        parts = text.split()
        if not parts:
            return _HELP
        cmd, args = parts[0].lower(), parts[1:]
        return {
            "add":      lambda: self._cmd_add(args),
            "budget":   lambda: self._cmd_budget(args),
            "report":   lambda: self._cmd_report(),
            "balance":  lambda: self._cmd_balance(),
            "compound": lambda: self._cmd_compound(args),
            "loan":     lambda: self._cmd_loan(args),
            "roi":      lambda: self._cmd_roi(args),
            "tax":      lambda: self._cmd_tax(args),
            "tip":      lambda: self._cmd_tip(args),
        }.get(cmd, lambda: _HELP if cmd in ("help", "") else f"Unknown command: `{cmd}`. Type `help`.")()

    # ── add ────────────────────────────────────────────────────────────────
    def _cmd_add(self, args: list[str]) -> str:
        if not args:
            return "Usage: `add <amount> [category] [note]`"
        try:
            amount = float(args[0].lstrip("$€£¥+"))
            if amount <= 0:
                return "Amount must be positive."
        except ValueError:
            return f"Invalid amount: `{args[0]}`"
        category = args[1].lower() if len(args) > 1 else "misc"
        note     = " ".join(args[2:]) if len(args) > 2 else ""
        self._expenses.append({"amount": amount, "category": category, "note": note, "timestamp": time.time()})
        header = f"Logged **${amount:.2f}** → `{category}`" + (f" — _{note}_" if note else "")
        if self._budget > 0:
            spent = sum(e["amount"] for e in self._expenses)
            pct   = min(spent / self._budget * 100, 999)
            icon  = "🚨" if pct >= 100 else ("⚠" if pct >= 80 else "✓")
            return f"{header}\n{icon} Budget: ${spent:.2f} / ${self._budget:.2f} ({pct:.0f}%)"
        return header

    # ── budget ─────────────────────────────────────────────────────────────
    def _cmd_budget(self, args: list[str]) -> str:
        if not args:
            return "Usage: `budget <amount>`"
        try:
            amount = float(args[0].lstrip("$€£¥"))
        except ValueError:
            return f"Invalid amount: `{args[0]}`"
        self._budget = amount
        spent = sum(e["amount"] for e in self._expenses)
        pct   = min(spent / amount * 100, 999) if amount > 0 else 0
        return f"Budget set: **${amount:.2f}** | Spent so far: ${spent:.2f} ({pct:.0f}%)"

    # ── report ─────────────────────────────────────────────────────────────
    def _cmd_report(self) -> str:
        if not self._expenses:
            return "No expenses yet. Try: `add 25 food coffee`"
        by_cat: dict[str, float] = {}
        for e in self._expenses:
            by_cat[e["category"]] = by_cat.get(e["category"], 0) + e["amount"]
        total = sum(by_cat.values())
        rows  = [f"  **{cat}**: ${amt:.2f} ({amt/total*100:.0f}%)" for cat, amt in sorted(by_cat.items(), key=lambda x: -x[1])]
        return f"**Expense Report** ({len(self._expenses)} transactions)\n\n" + "\n".join(rows) + f"\n\n**Total: ${total:.2f}**"

    # ── balance ────────────────────────────────────────────────────────────
    def _cmd_balance(self) -> str:
        spent = sum(e["amount"] for e in self._expenses)
        if self._budget <= 0:
            return f"**Balance**\n\nTotal spent: **${spent:.2f}** ({len(self._expenses)} transactions)\nBudget: _not set_"
        rem  = self._budget - spent
        pct  = min(spent / self._budget * 100, 999)
        icon = "🚨" if pct >= 100 else ("⚠" if pct >= 80 else "✓")
        return (
            f"**Monthly Budget Balance**\n\n"
            f"{icon} ${spent:.2f} / ${self._budget:.2f} ({pct:.0f}%)\n"
            f"{'${:.2f} left'.format(rem) if rem >= 0 else '${:.2f} over budget'.format(-rem)}"
        )

    # ── compound ───────────────────────────────────────────────────────────
    def _cmd_compound(self, args: list[str]) -> str:
        if len(args) < 3:
            return "Usage: `compound <principal> <rate%> <years>`  e.g. `compound 10000 7 20`"
        try:
            p, r, t = float(args[0].lstrip("$")), float(args[1].rstrip("%")) / 100, float(args[2])
        except ValueError:
            return "Invalid numbers."
        n  = 12.0
        fv = p * (1 + r / n) ** (n * t)
        return f"**Compound Interest**\n\nPrincipal: ${p:,.2f} | Rate: {r*100:.2f}% | Term: {t:.0f}y\n→ **Future Value: ${fv:,.2f}** (interest: ${fv-p:,.2f})"

    # ── loan ───────────────────────────────────────────────────────────────
    def _cmd_loan(self, args: list[str]) -> str:
        if len(args) < 3:
            return "Usage: `loan <principal> <rate%> <months>`  e.g. `loan 20000 5.5 60`"
        try:
            p, annual, n = float(args[0].lstrip("$")), float(args[1].rstrip("%")), float(args[2])
        except ValueError:
            return "Invalid numbers."
        r = annual / 100 / 12
        monthly = (p * r * (1 + r) ** n / ((1 + r) ** n - 1)) if r else p / n
        total   = monthly * n
        return f"**Loan Calculator**\n\nPrincipal: ${p:,.2f} | Rate: {annual:.2f}% | Term: {n:.0f}mo\n→ **Monthly: ${monthly:,.2f}** | Total: ${total:,.2f} | Interest: ${total-p:,.2f}"

    # ── roi ────────────────────────────────────────────────────────────────
    def _cmd_roi(self, args: list[str]) -> str:
        if len(args) < 2:
            return "Usage: `roi <investment> <return>`  e.g. `roi 5000 7500`"
        try:
            inv, ret = float(args[0].lstrip("$")), float(args[1].lstrip("$"))
        except ValueError:
            return "Invalid numbers."
        if inv == 0:
            return "Investment cannot be zero."
        gain = ret - inv
        roi  = gain / inv * 100
        return f"**ROI**\n\nInvestment: ${inv:,.2f} | Return: ${ret:,.2f}\n→ Gain: ${gain:+,.2f} | **ROI: {roi:+.2f}%**"

    # ── tax ────────────────────────────────────────────────────────────────
    def _cmd_tax(self, args: list[str]) -> str:
        if not args:
            return "Usage: `tax <income> [bracket%]`  e.g. `tax 75000 25`"
        try:
            income = float(args[0].lstrip("$"))
            rate   = float(args[1].rstrip("%")) if len(args) > 1 else 25.0
        except ValueError:
            return "Invalid numbers."
        tax = income * rate / 100
        return f"**Tax Estimate**\n\nIncome: ${income:,.2f} | Rate: {rate:.1f}%\n→ Tax: **${tax:,.2f}** | Net: **${income-tax:,.2f}**\n_Simplified estimate — consult a professional._"

    # ── tip ────────────────────────────────────────────────────────────────
    def _cmd_tip(self, args: list[str]) -> str:
        if not args:
            return "Usage: `tip <bill> [percent]`  e.g. `tip 45 20`"
        try:
            bill = float(args[0].lstrip("$"))
            pct  = float(args[1].rstrip("%")) if len(args) > 1 else 15.0
        except ValueError:
            return "Invalid numbers."
        tip   = bill * pct / 100
        total = bill + tip
        splits = "\n".join(f"  {n} people: ${total/n:.2f}/person" for n in (1, 2, 3, 4, 5))
        return f"**Tip Calculator**\n\nBill: ${bill:.2f} | Tip ({pct:.0f}%): ${tip:.2f} | **Total: ${total:.2f}**\n\n{splits}"

    def _current_task_description(self) -> str:
        return f"financier — {len(self._expenses)} expenses"
